File: packages/shared-tags/src/record-tag-event.ts
/**
 * iter30 WS2 — `recordTagEvent()` idempotent write contract.
 *
 * Spec: `.planning/iter30/ws-2-7-detail.md` §3 (Option C: TS+Python port).
 *
 * Design points:
 *  - Idempotency key: sha256(source|sourceDocId|rawText|kind|id).slice(0,32).
 *    Same content from same source for same entity → same eventId; the
 *    second create raises `ALREADY_EXISTS` and we return `created:false`.
 *  - Validation: missing both `userId` and `jobId` (when caller uses the
 *    PA-shorthand input shape) → throw. Empty `entityRef.id` → throw.
 *  - rawText > 500 chars → truncated, with a `truncated:true` debug flag
 *    on the result. This mirrors detail-plan §3.3.
 *  - Firestore write uses `.create()` so two concurrent writers for the
 *    same key collide deterministically; no read-then-write race.
 *
 * Wave-1 surface: this module ships the shape + idempotency contract.
 * Wave-2 will add JSON-Schema emit + GH Packages publish; Phase-2 ships
 * the worker that consumes events.
 */

import type {
  RecordTagEventInput,
  TagEvent,
} from "./schemas.js"
import {
  RecordTagEventInputSchema,
  TagEventSchema,
} from "./schemas.js"
import {
  TAG_EVENT_SCHEMA_VERSION,
  SHARED_TAG_COLLECTIONS,
  type TagEventSource,
  type TagType,
  type EntityRef,
  type EntityKind,
} from "./types.js"
import { sha256Hex } from "./sha256.js"

// ─────────────────────────────────────────────────────────────────
// Public types
// ─────────────────────────────────────────────────────────────────

/**
 * PA-shorthand input shape from the Day-1 brief. Either `userId` or
 * `jobId` must be set (xor); resolves to an `EntityRef`. Callers that
 * need the richer cross-repo shape (e.g. `scraping-github-repo:foo/bar`)
 * use `recordTagEventWithEntityRef()` directly.
 */
export interface RecordTagEventArgs {
  userId?: string
  jobId?: string
  rawTag: string
  source: TagEventSource
  /** Source-specific dot-path for audit. Defaults to "rawTag" if omitted. */
  sourceField?: string
  /** Source doc id; defaults to userId or jobId when omitted. */
  sourceDocId?: string
  type?: TagType
  evidence?: string
}

export interface RecordTagEventResult {
  eventId: string
  /** True if the doc was created by this call. False on duplicate. */
  created: boolean
  /** Compatibility alias for callers that prefer `idempotent` semantics. */
  idempotent: boolean
  reason?: "duplicate" | "validation-error" | "truncated"
  /** True if rawTag was truncated to fit the 500-char cap. */
  truncated?: boolean
}

/**
 * Minimal Firestore surface we depend on. Kept narrow so this package can
 * accept either `firebase-admin/firestore.Firestore` or the modular
 * `@google-cloud/firestore` client without pulling either as a hard dep.
 */
export interface FirestoreLike {
  collection(name: string): FirestoreCollectionLike
}
export interface FirestoreCollectionLike {
  doc(id: string): FirestoreDocLike
}
export interface FirestoreDocLike {
  create(data: Record<string, unknown>): Promise<unknown>
}

export interface RecordTagEventDeps {
  firestore: FirestoreLike
  /** Override for tests. Defaults to `() => new Date().toISOString()`. */
  now?: () => string
}

// ─────────────────────────────────────────────────────────────────
// Idempotency key
// ─────────────────────────────────────────────────────────────────

/**
 * Build the canonical idempotency input. Lowercased + trimmed `rawText`
 * collapses "ML"/"ml"/" ML " to one event. Exported for the Python port
 * + parity tests.
 */
export function buildIdempotencyInput(args: {
  source: TagEventSource
  sourceDocId: string
  rawText: string
  entityRef: EntityRef
}): string {
  return [
    args.source,
    args.sourceDocId,
    args.rawText.toLowerCase().trim(),
    args.entityRef.kind,
    args.entityRef.id,
  ].join("|")
}

export function computeEventId(args: {
  source: TagEventSource
  sourceDocId: string
  rawText: string
  entityRef: EntityRef
}): string {
  return sha256Hex(buildIdempotencyInput(args)).slice(0, 32)
}

// ─────────────────────────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────────────────────────

const RAW_TEXT_MAX = 500

/**
 * Resolve PA-shorthand `{userId, jobId}` → `EntityRef`. Exactly one of
 * `userId` or `jobId` must be set.
 */
export function resolveEntityRef(args: {
  userId?: string
  jobId?: string
}): EntityRef {
  const hasUser = typeof args.userId === "string" && args.userId.length > 0
  const hasJob = typeof args.jobId === "string" && args.jobId.length > 0
  if (hasUser === hasJob) {
    throw new Error(
      "[shared-tags] recordTagEvent: exactly one of `userId` or `jobId` must be set",
    )
  }
  if (hasUser) {
    return { kind: "pa-user" satisfies EntityKind, id: args.userId! }
  }
  return { kind: "pa-job" satisfies EntityKind, id: args.jobId! }
}

function truncateRawText(input: string): { rawText: string; truncated: boolean } {
  if (input.length <= RAW_TEXT_MAX) return { rawText: input, truncated: false }
  return { rawText: input.slice(0, RAW_TEXT_MAX), truncated: true }
}

function isAlreadyExists(err: unknown): boolean {
  if (!err || typeof err !== "object") return false
  // firebase-admin/firestore: throws with `code === 6` (ALREADY_EXISTS gRPC)
  // or `code === "already-exists"` depending on transport.
  const e = err as { code?: number | string; message?: string }
  if (e.code === 6) return true
  if (e.code === "already-exists") return true
  if (e.code === "ALREADY_EXISTS") return true
  if (typeof e.message === "string" && /ALREADY_EXISTS|already exists/i.test(e.message)) {
    return true
  }
  return false
}

// ─────────────────────────────────────────────────────────────────
// Main entry: PA-shorthand
// ─────────────────────────────────────────────────────────────────

/**
 * Idempotent append to `pa-tag-events`. Returns the eventId; caller can
 * trust that two calls with byte-identical input always produce the same
 * eventId and at most one Firestore document.
 *
 * Validation errors (missing entity, empty rawTag, etc.) THROW — caller
 * is expected to wrap in fire-and-forget try/catch on hot paths
 * (realtime-tagger, cv-ingest), per detail-plan §3.3.
 */
export async function recordTagEvent(
  args: RecordTagEventArgs,
  deps: RecordTagEventDeps,
): Promise<RecordTagEventResult> {
  if (typeof args.rawTag !== "string" || args.rawTag.trim().length === 0) {
    throw new Error("[shared-tags] recordTagEvent: rawTag must be a non-empty string")
  }
  const entityRef = resolveEntityRef({
    userId: args.userId,
    jobId: args.jobId,
  })

  const { rawText, truncated } = truncateRawText(args.rawTag)
  const sourceDocId = args.sourceDocId ?? entityRef.id
  const sourceField = args.sourceField ?? "rawTag"
  const type: TagType = args.type ?? "skill"

  const input: RecordTagEventInput = {
    source: args.source,
    sourceDocId,
    sourceField,
    rawText,
    type,
    entityRef,
    context: args.evidence,
  }

  const parsed = RecordTagEventInputSchema.safeParse(input)
  if (!parsed.success) {
    throw new Error(
      `[shared-tags] recordTagEvent: validation failed — ${parsed.error.message}`,
    )
  }

  return await recordTagEventWithEntityRef(parsed.data, deps, { truncated })
}

// ─────────────────────────────────────────────────────────────────
// Cross-repo entry: explicit EntityRef
// ─────────────────────────────────────────────────────────────────

/**
 * Lower-level entry point used by scraping pipelines that emit tags for
 * non-PA entities (`scraping-github-repo:foo/bar`, etc.). The PA path
 * resolves through here as well after PA-shorthand expansion.
 */
export async function recordTagEventWithEntityRef(
  input: RecordTagEventInput,
  deps: RecordTagEventDeps,
  meta: { truncated?: boolean } = {},
): Promise<RecordTagEventResult> {
  const eventId = computeEventId({
    source: input.source,
    sourceDocId: input.sourceDocId,
    rawText: input.rawText,
    entityRef: input.entityRef,
  })

  const now = (deps.now ?? (() => new Date().toISOString()))()

  const event: TagEvent = TagEventSchema.parse({
    id: eventId,
    source: input.source,
    sourceDocId: input.sourceDocId,
    sourceField: input.sourceField,
    rawText: input.rawText,
    type: input.type,
    entityRef: input.entityRef,
    context: input.context,
    status: "pending",
    normalizedTo: null,
    decision: null,
    workerConfidence: null,
    schemaVersion: TAG_EVENT_SCHEMA_VERSION,
    createdAt: now,
    processedAt: null,
    processingDurationMs: null,
  })

  const ref = deps.firestore
    .collection(SHARED_TAG_COLLECTIONS.tagEvents)
    .doc(eventId)

  try {
    await ref.create(event as unknown as Record<string, unknown>)
    return {
      eventId,
      created: true,
      idempotent: false,
      ...(meta.truncated ? { truncated: true, reason: "truncated" as const } : {}),
    }
  } catch (err) {
    if (isAlreadyExists(err)) {
      return {
        eventId,
        created: false,
        idempotent: true,
        reason: "duplicate",
        ...(meta.truncated ? { truncated: true } : {}),
      }
    }
    throw err
  }
}

File: packages/shared-tags-py/wekruit_shared_tags/record_tag_event.py
"""
iter30 WS2 — Python port of `recordTagEvent()` idempotent write contract.

Mirrors `packages/shared-tags/src/record-tag-event.ts`. Adam-locked
constraints:

  - Idempotency-key BYTE-PARITY with TS:
      sha256("{source}|{sourceDocId}|{rawText.lower().strip()}|{kind}|{id}")
      .hexdigest()[:32]
  - Validation: missing both userId+jobId → ValueError. Empty rawTag →
    ValueError. rawText > 500 chars → truncate + result.truncated=True.
  - Firestore `.create()` semantics: ALREADY_EXISTS → return
    {created: False, idempotent: True, reason: "duplicate"}; other
    errors propagate.
  - Async + sync Firestore clients both supported (see `_maybe_await`).

Caller surface:

    from wekruit_shared_tags import record_tag_event
    result = await record_tag_event(
        {"userId": "u1", "rawTag": "ML", "source": "scraping-github"},
        firestore=fs,
    )
"""
from __future__ import annotations

import asyncio
import inspect
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Mapping, TypedDict

from .schemas import EntityRef, RecordTagEventInput, TagEvent
from .sha256 import sha256_hex
from .types import (
    SHARED_TAG_COLLECTIONS,
    TAG_EVENT_SCHEMA_VERSION,
    EntityKind,
    TagEventSource,
    TagType,
)

# ─────────────────────────────────────────────────────────────────
# Public input/result shapes
# ─────────────────────────────────────────────────────────────────


class RecordTagEventArgs(TypedDict, total=False):
    """
    PA-shorthand input. Either `userId` or `jobId` must be set (xor).
    `type` defaults to "skill" (same as TS). `sourceField` defaults to
    "rawTag", `sourceDocId` defaults to userId/jobId.
    """

    userId: str
    jobId: str
    rawTag: str
    source: TagEventSource
    sourceField: str
    sourceDocId: str
    type: TagType
    evidence: str


@dataclass
class RecordTagEventResult:
    """
    Return shape — mirrors TS `RecordTagEventResult`. `idempotent` is the
    semantic-flip of `created` for caller convenience.
    """

    eventId: str
    created: bool
    idempotent: bool
    reason: str | None = None
    truncated: bool = False

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "eventId": self.eventId,
            "created": self.created,
            "idempotent": self.idempotent,
        }
        if self.reason is not None:
            out["reason"] = self.reason
        if self.truncated:
            out["truncated"] = True
        return out


# ─────────────────────────────────────────────────────────────────
# Idempotency key
# ─────────────────────────────────────────────────────────────────


def build_idempotency_input(
    *,
    source: str,
    sourceDocId: str,
    rawText: str,
    entityRef: EntityRef | Mapping[str, str],
) -> str:
    """
    Build the canonical pipe-joined idempotency string. BYTE-PARITY with
    TS `buildIdempotencyInput()` is mandatory — see the TS test fixture
    `pa-realtime-tagger|msg-1|ml|pa-user|usr-abc`.
    """
    if isinstance(entityRef, EntityRef):
        kind = entityRef.kind
        ref_id = entityRef.id
    else:
        kind = entityRef["kind"]
        ref_id = entityRef["id"]
    return "|".join(
        [
            source,
            sourceDocId,
            rawText.lower().strip(),
            kind,
            ref_id,
        ]
    )


def compute_event_id(
    *,
    source: str,
    sourceDocId: str,
    rawText: str,
    entityRef: EntityRef | Mapping[str, str],
) -> str:
    """sha256 prefix; 32 hex chars; matches TS `computeEventId`."""
    return sha256_hex(
        build_idempotency_input(
            source=source,
            sourceDocId=sourceDocId,
            rawText=rawText,
            entityRef=entityRef,
        )
    )[:32]


# ─────────────────────────────────────────────────────────────────
# Internal helpers
# ─────────────────────────────────────────────────────────────────

_RAW_TEXT_MAX = 500


def _truncate_raw_text(s: str) -> tuple[str, bool]:
    if len(s) <= _RAW_TEXT_MAX:
        return s, False
    return s[:_RAW_TEXT_MAX], True


def _resolve_entity_ref(*, userId: str | None, jobId: str | None) -> EntityRef:
    has_user = isinstance(userId, str) and len(userId) > 0
    has_job = isinstance(jobId, str) and len(jobId) > 0
    if has_user == has_job:
        raise ValueError(
            "[shared-tags] record_tag_event: exactly one of `userId` or `jobId` must be set"
        )
    if has_user:
        return EntityRef(kind="pa-user", id=userId)  # type: ignore[arg-type]
    return EntityRef(kind="pa-job", id=jobId)  # type: ignore[arg-type]


def _is_already_exists(err: BaseException) -> bool:
    """
    Detect Firestore ALREADY_EXISTS across SDK transports.

    google-cloud-firestore raises `google.api_core.exceptions.AlreadyExists`
    (gRPC code=6). Plain duck-typed fakes may raise anything with a
    matching attribute or message. Match all of them.
    """
    code = getattr(err, "code", None)
    # google.api_core ALREADY_EXISTS has grpc_status_code = 6
    if isinstance(code, int) and code == 6:
        return True
    if isinstance(code, str) and code.lower().replace("_", "-") in (
        "already-exists",
    ):
        return True
    # gRPC status object
    grpc_status = getattr(err, "grpc_status_code", None)
    if grpc_status is not None and getattr(grpc_status, "value", None) == 6:
        return True
    cls_name = type(err).__name__
    if cls_name in ("AlreadyExists", "AlreadyExistsError"):
        return True
    msg = str(err)
    if "ALREADY_EXISTS" in msg or "already exists" in msg.lower():
        return True
    return False


async def _maybe_await(value: Any) -> Any:
    """Accept both sync and async Firestore clients."""
    if inspect.isawaitable(value):
        return await value
    return value


def _default_now() -> str:
    # ISO-8601 UTC with 'Z' suffix (matches TS `new Date().toISOString()`)
    now = datetime.now(timezone.utc)
    # millisecond precision, drop microseconds to avoid drift in fixtures
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


# ─────────────────────────────────────────────────────────────────
# Public entry — PA-shorthand
# ─────────────────────────────────────────────────────────────────


async def record_tag_event(
    args: RecordTagEventArgs | Mapping[str, Any],
    *,
    firestore: Any,
    now: Callable[[], str] | None = None,
) -> RecordTagEventResult:
    """
    Idempotent append to `pa-tag-events`. Two calls with byte-identical
    input produce the same eventId and at most one Firestore document.

    Validation errors raise `ValueError` — caller is expected to wrap in
    fire-and-forget try/except on hot paths (matches TS contract §3.3).
    """
    raw_tag = args.get("rawTag")
    if not isinstance(raw_tag, str) or len(raw_tag.strip()) == 0:
        raise ValueError(
            "[shared-tags] record_tag_event: rawTag must be a non-empty string"
        )

    entity_ref = _resolve_entity_ref(
        userId=args.get("userId"),
        jobId=args.get("jobId"),
    )

    raw_text, truncated = _truncate_raw_text(raw_tag)
    source = args.get("source")
    if source is None:
        raise ValueError("[shared-tags] record_tag_event: `source` is required")

    source_doc_id = args.get("sourceDocId") or entity_ref.id
    source_field = args.get("sourceField") or "rawTag"
    tag_type: TagType = args.get("type") or "skill"  # type: ignore[assignment]
    context = args.get("evidence")

    # Validate via pydantic (Zod-strict parity)
    parsed = RecordTagEventInput(
        source=source,
        sourceDocId=source_doc_id,
        sourceField=source_field,
        rawText=raw_text,
        type=tag_type,
        entityRef=entity_ref,
        context=context,
    )

    return await record_tag_event_with_entity_ref(
        parsed,
        firestore=firestore,
        now=now,
        truncated=truncated,
    )


# ─────────────────────────────────────────────────────────────────
# Public entry — explicit EntityRef (cross-repo)
# ─────────────────────────────────────────────────────────────────


async def record_tag_event_with_entity_ref(
    input: RecordTagEventInput | Mapping[str, Any],
    *,
    firestore: Any,
    now: Callable[[], str] | None = None,
    truncated: bool = False,
) -> RecordTagEventResult:
    """
    Lower-level entry used by scraping pipelines that emit tags for
    non-PA entities (e.g. `scraping-github-repo:foo/bar`).
    """
    if not isinstance(input, RecordTagEventInput):
        input = RecordTagEventInput.model_validate(dict(input))

    event_id = compute_event_id(
        source=input.source,
        sourceDocId=input.sourceDocId,
        rawText=input.rawText,
        entityRef=input.entityRef,
    )

    now_str = (now or _default_now)()

    event = TagEvent(
        id=event_id,
        source=input.source,
        sourceDocId=input.sourceDocId,
        sourceField=input.sourceField,
        rawText=input.rawText,
        type=input.type,
        entityRef=input.entityRef,
        context=input.context,
        status="pending",
        normalizedTo=None,
        decision=None,
        workerConfidence=None,
        schemaVersion=TAG_EVENT_SCHEMA_VERSION,
        createdAt=now_str,
        processedAt=None,
        processingDurationMs=None,
    )

    payload = event.model_dump(mode="json")

    collection = firestore.collection(SHARED_TAG_COLLECTIONS["tagEvents"])
    # google-cloud-firestore Python SDK uses `.document(id)`. Some test
    # fakes mirror TS `.doc(id)` — accept both.
    if hasattr(collection, "document"):
        doc_ref = collection.document(event_id)
    else:
        doc_ref = collection.doc(event_id)  # type: ignore[attr-defined]

    try:
        await _maybe_await(doc_ref.create(payload))
        return RecordTagEventResult(
            eventId=event_id,
            created=True,
            idempotent=False,
            reason="truncated" if truncated else None,
            truncated=truncated,
        )
    except Exception as err:  # noqa: BLE001 — we re-raise non-AlreadyExists
        if _is_already_exists(err):
            return RecordTagEventResult(
                eventId=event_id,
                created=False,
                idempotent=True,
                reason="duplicate",
                truncated=truncated,
            )
        raise


__all__ = [
    "RecordTagEventArgs",
    "RecordTagEventResult",
    "build_idempotency_input",
    "compute_event_id",
    "record_tag_event",
    "record_tag_event_with_entity_ref",
]
